internal/client/db/utils.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"math"
+
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Meta column. The set of keys in the column is shared by full && light storages.
+const columnMeta = columns.Meta
+
+type finalizedState[H, N any] struct {
+	Hash   H
+	Number N
+}
+
+// Database metadata.
+type meta[H, N any] struct {
+	// Hash of the best known block.
+	BestHash H
+	// Number of the best known block.
+	BestNumber N
+	// Hash of the best finalized block.
+	FinalizedHash H
+	// Number of the best finalized block.
+	FinalizedNumber N
+	// Hash of the genesis block.
+	GenesisHash H
+	// Finalized state, if any
+	FinalizedState *finalizedState[H, N]
+	// Block gap, start and end inclusive, if any.
+	BlockGap *[2]N
+}
+
+// A block lookup key: used for canonical lookup from block number to hash
+type numberIndexKey [4]byte
+
+// Convert block number into short lookup key (LE representation) for
+// blocks that are in the canonical chain.
+//
+// In the current database schema, this kind of key is only used for
+// lookups into an index, NOT for storing header data or others.
+func newNumberIndexKey[N runtime.Number](num N) (numberIndexKey, error) {
+	if num > N(math.MaxUint32) {
+		return numberIndexKey{}, fmt.Errorf("block number cannot be converted to uint32")
+	}
+	var n uint32 = uint32(num)
+
+	return numberIndexKey{byte(n >> 24), byte((n >> 16) & 0xff), byte((n >> 8) & 0xff), byte(n & 0xff)}, nil
```

Comment:
Could we try to clean this up a bit, It is rather hard to read IMO. Maybe make variables describing what each component is, or at least add a detailed comment explaining what is happening

Reply:
I copied this from substrate. This is just little endian representation into a fixed size array of length 4.  I think the accompanying tests make this pretty clear.